aws-lc-rs/src/ec/key_pair.rs

```diff
@@ -173,16 +188,73 @@ impl EcdsaKeyPair {
     ///
     #[inline]
     pub fn sign(&self, _rng: &dyn SecureRandom, message: &[u8]) -> Result<Signature, Unspecified> {
-        unsafe {
-            let digest = digest::digest(self.algorithm.digest, message);
-            let digest = digest.as_ref();
-            let ecdsa_sig = LcPtr::new(ECDSA_do_sign(digest.as_ptr(), digest.len(), *self.ec_key))?;
-            match self.algorithm.sig_format {
-                EcdsaSignatureFormat::ASN1 => ec::ecdsa_sig_to_asn1(&ecdsa_sig),
-                EcdsaSignatureFormat::Fixed => {
-                    ec::ecdsa_sig_to_fixed(self.algorithm.id, &ecdsa_sig)
-                }
-            }
-        }
+        let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) })
+            .map_err(|_| Unspecified)?;
+
+        if 1 != unsafe { EVP_PKEY_sign_init(*pkey_ctx) } {
+            return Err(Unspecified);
+        };
+
+        let mut context = digest::digest_ctx::DigestContext::new(self.algorithm.digest)?;
+        unsafe { EVP_MD_CTX_set_pkey_ctx(context.as_mut_ptr(), *pkey_ctx) };
+
+        let mut out_sig = vec![0u8; get_signature_length(&mut context)?];
+
+        let out_sig = compute_ecdsa_signature(&mut context, message, out_sig.as_mut_slice())?;
+
+        Ok(match self.algorithm.sig_format {
+            EcdsaSignatureFormat::ASN1 => Signature::new(|slice| {
+                slice[..out_sig.len()].copy_from_slice(out_sig);
+                out_sig.len()
+            }),
+            EcdsaSignatureFormat::Fixed => ec::ecdsa_asn1_to_fixed(self.algorithm.id, out_sig)?,
+        })
+    }
+}
+
+#[inline]
+fn get_signature_length(ctx: &mut DigestContext) -> Result<usize, Unspecified> {
+    let mut out_sig_len = MaybeUninit::<usize>::uninit();
+
+    // determine signature size
+    let result = unsafe {
+        EVP_DigestSign(
+            ctx.as_mut_ptr(),
+            null_mut(),
+            out_sig_len.as_mut_ptr(),
+            null(),
+            0,
+        )
+    };
+    if 1 != result {
+        return Err(Unspecified);
+    }
+
+    Ok(unsafe { out_sig_len.assume_init() })
+}
+
+#[inline]
+fn compute_ecdsa_signature<'a>(
+    ctx: &mut DigestContext,
+    message: &[u8],
+    signature: &'a mut [u8],
+) -> Result<&'a mut [u8], Unspecified> {
+    let mut out_sig_len = MaybeUninit::<usize>::new(signature.len());
+
+    let result = unsafe {
+        EVP_DigestSign(
+            ctx.as_mut_ptr(),
+            signature.as_mut_ptr(),
+            out_sig_len.as_mut_ptr(),
+            message.as_ptr(),
+            message.len(),
+        )
+    };
+    if 1 != result {
+        return Err(Unspecified);
     }
```

Comment:
There's no need for a `result` -- other places apply the condition directly to the invocation:
```
    if 1 != unsafe {
        EVP_DigestSign(
            ctx.as_mut_ptr(),
            signature.as_mut_ptr(),
            out_sig_len.as_mut_ptr(),
            message.as_ptr(),
            message.len(),
        ) {
        return Err(Unspecified);
    }
```